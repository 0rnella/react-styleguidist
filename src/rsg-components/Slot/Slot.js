// Inspired by https://github.com/camwest/react-slot-fill

import React from 'react';
import PropTypes from 'prop-types';

export default function Slot({ name, active, onlyActive, className, props = {} }, { slots }) {
	const fills = slots[name];
	if (!fills) {
		throw new Error(`Slot "${name}" not found, available slots: ${Object.keys(slots).join(', ')}`);
	}

	// { id: 'pizza', render: ({ foo }) => <div>{foo}</div> }
	const rendered = fills.map(({ id, render }, index) => {
		// Render only specified fill
		if (onlyActive && id !== active) {
			return null;
		}

		const { onClick } = props;
		props = {
			...props,
			id,
			// Set active prop to active fill
			active: active && id === active,
			// Pass fill ID to onClick event handler
			// eslint-disable-next-line react/prop-types
			onClick: onClick && ((...attrs) => onClick(id, ...attrs)),
		};

		const Component = render;
		return <Component key={index} {...props} />;
	});

	const filtered = rendered.filter(Boolean);
	if (filtered.length === 0) {
		return null;
	}

	return <div className={className}>{filtered}</div>;
}

Slot.propTypes = {
	name: PropTypes.string.isRequired,
	active: PropTypes.string,
	onlyActive: PropTypes.bool,
	props: PropTypes.object,
	className: PropTypes.string,
};
Slot.contextTypes = {
	slots: PropTypes.object.isRequired,
};
