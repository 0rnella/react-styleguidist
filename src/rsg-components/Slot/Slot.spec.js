import React from 'react';
import Slot from './Slot';

/* eslint-disable react/prop-types */

const Button = ({ onClick }) => <button onClick={onClick}>1</button>;
const Button2 = () => <button>2</button>;
const options = {
	context: {
		slots: {
			slot: [
				{
					id: 'one',
					render: Button,
				},
				{
					id: 'two',
					render: Button2,
				},
			],
		},
	},
};

it('should renderer slots and pass props', () => {
	const actual = shallow(<Slot name="slot" props={{ id: 'Pizza' }} />, options);

	expect(actual).toMatchSnapshot();
});

it('should pass active flag to active slot', () => {
	const actual = shallow(<Slot name="slot" active="two" />, options);

	expect(actual).toMatchSnapshot();
});

it('should renderer only active slot if onlyActive=true', () => {
	const actual = shallow(<Slot name="slot" active="two" onlyActive />, options);

	expect(actual).toMatchSnapshot();
});

it('should pass slot ID to onClick handler', () => {
	const onClick = jest.fn();
	const actual = mount(<Slot name="slot" props={{ onClick }} />, options);

	actual.find('button').first().simulate('click');

	expect(onClick).toBeCalledWith('one', expect.any(Object));
});

it('should return null if all slots render null', () => {
	const actual = render(<Slot name="slot" props={{ id: 'Pizza' }} />, {
		context: {
			slots: {
				slot: [
					{
						id: 'nope',
						type: 'slot',
						render: () => null,
					},
				],
			},
		},
	});

	expect(actual.node).toBeFalsy();
});
