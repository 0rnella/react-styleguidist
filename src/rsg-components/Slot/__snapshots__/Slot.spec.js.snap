// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass active flag to active slot 1`] = `
<div>
  <Button
    active={false}
    id="one"
  />
  <Button2
    active={true}
    id="two"
  />
</div>
`;

exports[`should renderer only active slot if onlyActive=true 1`] = `
<div>
  <Button2
    active={true}
    id="two"
  />
</div>
`;

exports[`should renderer slots and pass props 1`] = `
<div>
  <Button
    id="one"
  />
  <Button2
    id="two"
  />
</div>
`;
